mountpoint-s3/tests/fuse_tests/cache_test.rs

```diff
@@ -0,0 +1,121 @@
+use crate::common::fuse::{self, TestSessionConfig};
+use crate::common::s3::{get_express_cache_bucket, get_test_bucket_and_prefix};
+use mountpoint_s3::data_cache::{DataCache, DiskDataCache, DiskDataCacheConfig, ExpressDataCache};
+use mountpoint_s3_client::S3CrtClient;
+use rand::{Rng, RngCore, SeedableRng};
+use rand_chacha::ChaChaRng;
+use std::fs;
+use std::path::PathBuf;
+use std::thread::sleep;
+use std::time::Duration;
+use test_case::test_case;
+
+#[cfg(all(feature = "s3_tests", feature = "s3express_tests"))]
+#[test_case("key", 100, 1024; "simple")]
+#[test_case("£", 100, 1024; "non-ascii key")]
+#[test_case("key", 1024, 1024; "long key")]
+#[test_case("key", 100, 1024 * 1024; "big file")]
+fn express_cache_write_read(key_suffix: &str, key_size: usize, object_size: usize) {
+    cache_write_read_base(
+        key_suffix,
+        key_size,
+        object_size,
+        express_cache_factory,
+        "express_cache_write_read",
+    )
+}
+
+#[cfg(feature = "s3_tests")]
+#[test_case("key", 100, 1024; "simple")]
+#[test_case("£", 100, 1024; "non-ascii key")]
+#[test_case("key", 1024, 1024; "long key")]
+#[test_case("key", 100, 1024 * 1024; "big file")]
+fn disk_cache_write_read(key_suffix: &str, key_size: usize, object_size: usize) {
+    let cache_dir = tempfile::tempdir().unwrap();
+    cache_write_read_base(
+        key_suffix,
+        key_size,
+        object_size,
+        disk_cache_factory(cache_dir.path().to_owned()),
+        "disk_cache_write_read",
+    );
+}
+
+fn cache_write_read_base<Cache, CacheFactory>(
+    key_suffix: &str,
+    key_size: usize,
+    object_size: usize,
+    cache_factory: CacheFactory,
+    test_name: &str,
+) where
+    Cache: DataCache + Send + Sync + 'static,
+    CacheFactory: FnOnce(S3CrtClient, u64) -> Cache,
+{
+    // mount a bucket
+    const BLOCK_SIZE: u64 = 512 * 1024;
+    let mut test_config: TestSessionConfig = TestSessionConfig::default();
+    test_config.filesystem_config.cache_config.serve_lookup_from_cache = true;
+    test_config.filesystem_config.cache_config.dir_ttl = Duration::from_secs(3600);
+    test_config.filesystem_config.cache_config.file_ttl = Duration::from_secs(3600);
+    let (_, prefix) = get_test_bucket_and_prefix(test_name);
+    let (mount_point, _session, mut client) =
+        fuse::s3_session::new_with_cache_factory(prefix.clone(), cache_factory, test_config, BLOCK_SIZE);
+
+    // write an object, no caching happens yet
+    let key = get_object_key(&prefix, key_suffix, key_size);
+    let path = mount_point.path().join(&key);
+    let written = random_binary_data(object_size);
+    fs::write(&path, &written).expect("write should succeed");
+
+    // first read should be from the source bucket and be cached
+    let read = fs::read(&path).expect("read should succeed");
+    assert_eq!(read, written);
+
+    // cache population is async, 1 second should be enough for it to finish
```

Comment:
Is there a better way for us to tell if it's done?

Reply:
I'm thinking that avoiding flakey tests is something we should do here. Is it possible to make things public for tests only? Otherwise, I wonder if it's possible for us to return some awaitable on cache write complete. 

Can this impact regular on-disk cache?

Reply:
another improvement which I can propose is to retry actual reads of the file until the object is served from cache (then now sleep after the first read will be required):
```rust
    // second read should be from the cache
    const MAX_WAIT_DURATION: std::time::Duration = std::time::Duration::from_secs(10);
    let st = std::time::Instant::now();
    loop {
        if st.elapsed() > MAX_WAIT_DURATION {
            panic!("timeout on waiting for data being served from the cache")
        }
        if let Ok(read) = fs::read(&path) {
            assert_eq!(read, written);
            break;
        }
        sleep(Duration::from_millis(100));
    }
```

Lmk if this looks better.

Reply:
Replaced a single `sleep` with retrying read logic

Reply:
Reverted back to `sleep` after @passaro [comment](https://github.com/awslabs/mountpoint-s3/pull/1071#discussion_r1831432734)